Confidential-Storage-Arch.md

```diff
@@ -0,0 +1,124 @@
+  enforce too much constraints on the deployment environment.
+- **Storage Architecture for the Trusted Domain**. This documentation should only define storage architectures
+  for the trusted domain. It may also give advices/recommendations for storage architectures outside of the trusted
+  domain, but it should not define/enforce anything out of the trusted domain.
+- **Focus on Confidentiality, Integrity instead of Performance**. There are always possible ways to improve performance,
+  and always be possible to treat performance by cost. So it should focus on defining storage architecture to ensure
+  confidentiality and integrity instead of improving performance.
+
+## Assumptions
+
+- All data originated from outside of the trusted domain are untrusted.
+- With integrity check/decryption, untrusted data may be transformed/converted into trusted data.
+- Trusted data stored within the trust domain will always be trustable.  
+- It's legal to send integrity enforced/encrypted data to software data cache services provided by service providers.
+- The secrets to verify/decrypt untrusted data should be derived from the remote attestation/KMS service.
+- Trusted data will be always accessed through trusted filesystems. If users want to directly access data from raw block
+  devices or untrusted filesystems, they must take the responsibility to verify/decrypt those data.
+- Guest memory is a sort of precious resource. It should not be used to store large scale of ephemeral storage content.
+
+## Candidate Technologies and Solutions
+
+With goals and assumptions defined above, the task becomes to define a software stack to convert untrusted storage data
+originated outside of the trusted domain into trusted filesystems within the trusted domain.
+
+### Sources of Untrusted Storage Data
+
+Let's first enumerate the possible sources of untrusted storage data as below:
+- initrd (measured)
+- virtio-blk devices
+- virtio-scsi devices
+- virtio-pmem devices  
+- nvme devices
+- virtio-fs devices
+- virtio-9p devices
+- network attached filesystems, such as NFS, CephFS etc. There's a special case for network attached filesystems, it may
+  be trusted filesystem if it's connected to a trusted server through secure connections.
+- software services providing storage data, for example an image cache service integrated into the image service.  
+
+The storage subsystem may be composed up with multiple layers, so we may enforce confidentiality at different positions.
+Here confidentiality enforcement means decryption and/or integrity verification.
+
+### Enforce Confidentiality at The Data Source Layer
+
+The confidentiality may be enforced at the data source layer, such as at the service agents or the NFS clients.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Data Source Layer](images/storage-enforce-confidentiality-at-source.png)
+
```

Comment:
I have a few questions/comments about the diagram:

- Does `Comp` stand for `Component` ?
- What does `Guarder` mean here?
- What is a `Normal` filesystem or block device?
- Is the `Service Agent` the kata agent for example?
- It would be clearer to draw a box for the trusted domain, showing how and when the data is converted from untrusted to trusted.
-